In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt


In [2]:
#read the excel file on loan data into the dataframe
df =pd.read_excel('Copy of loan.xlsx')
df.head()

Loan_ID Gender Married Dependents     Education Self_Employed  \
0  LP001002   Male      No          0      Graduate            No   
1  LP001003   Male     Yes          1      Graduate            No   
2  LP001005   Male     Yes          0      Graduate           Yes   
3  LP001006   Male     Yes          0  Not Graduate            No   
4  LP001008   Male      No          0      Graduate            No   

   ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
0             5849                0.0         NaN             360.0   
1             4583             1508.0       128.0             360.0   
2             3000                0.0        66.0             360.0   
3             2583             2358.0       120.0             360.0   
4             6000                0.0       141.0             360.0   

   Credit_History Property_Area Loan_Status  
0             1.0         Urban           Y  
1             1.0         Rural           N  
2             1.0         Urban           Y  
3             1.0         Urban           Y  
4             1.0         Urban           Y

In [3]:
#data size?
df.shape

(614, 13)

So 12 variables, one output and 614 entries. THe dataset is quite small but we'll make do

In [4]:
#cursory exploration of the data
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 614 entries, 0 to 613
Data columns (total 13 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Loan_ID            614 non-null    object 
 1   Gender             601 non-null    object 
 2   Married            611 non-null    object 
 3   Dependents         599 non-null    object 
 4   Education          614 non-null    object 
 5   Self_Employed      582 non-null    object 
 6   ApplicantIncome    614 non-null    int64  
 7   CoapplicantIncome  614 non-null    float64
 8   LoanAmount         592 non-null    float64
 9   Loan_Amount_Term   600 non-null    float64
 10  Credit_History     564 non-null    float64
 11  Property_Area      614 non-null    object 
 12  Loan_Status        614 non-null    object 
dtypes: float64(4), int64(1), object(8)
memory usage: 62.5+ KB


In [5]:
#Let's examine the missing values
df.isnull().sum()

Loan_ID               0
Gender               13
Married               3
Dependents           15
Education             0
Self_Employed        32
ApplicantIncome       0
CoapplicantIncome     0
LoanAmount           22
Loan_Amount_Term     14
Credit_History       50
Property_Area         0
Loan_Status           0
dtype: int64

In [6]:
#lets take a look at the unique values for each column
df.nunique()

Loan_ID              614
Gender                 2
Married                2
Dependents             4
Education              2
Self_Employed          2
ApplicantIncome      505
CoapplicantIncome    287
LoanAmount           203
Loan_Amount_Term      10
Credit_History         2
Property_Area          3
Loan_Status            2
dtype: int64

<h2> Attending to the missing values in each data column </h2>

In [7]:
#Gender
#Looking for clues to guide replacing the missing data
df[df["Gender"].isnull()]

Loan_ID Gender Married Dependents     Education Self_Employed  \
23   LP001050    NaN     Yes          2  Not Graduate            No   
126  LP001448    NaN     Yes         3+      Graduate            No   
171  LP001585    NaN     Yes         3+      Graduate            No   
188  LP001644    NaN     Yes          0      Graduate           Yes   
314  LP002024    NaN     Yes          0      Graduate            No   
334  LP002103    NaN     Yes          1      Graduate           Yes   
460  LP002478    NaN     Yes          0      Graduate           Yes   
467  LP002501    NaN     Yes          0      Graduate            No   
477  LP002530    NaN     Yes          2      Graduate            No   
507  LP002625    NaN      No          0      Graduate            No   
576  LP002872    NaN     Yes          0      Graduate            No   
588  LP002925    NaN      No          0      Graduate            No   
592  LP002933    NaN      No         3+      Graduate           Yes   

     ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
23              3365             1917.0       112.0             360.0   
126            23803                0.0       370.0             360.0   
171            51763                0.0       700.0             300.0   
188              674             5296.0       168.0             360.0   
314             2473             1843.0       159.0             360.0   
334             9833             1833.0       182.0             180.0   
460             2083             4083.0       160.0             360.0   
467            16692                0.0       110.0             360.0   
477             2873             1872.0       132.0             360.0   
507             3583                0.0        96.0             360.0   
576             3087             2210.0       136.0             360.0   
588             4750                0.0        94.0             360.0   
592             9357                0.0       292.0             360.0   

     Credit_History Property_Area Loan_Status  
23              0.0         Rural           N  
126             1.0         Rural           Y  
171             1.0         Urban           Y  
188             1.0         Rural           Y  
314             1.0         Rural           N  
334             1.0         Urban           Y  
460             NaN     Semiurban           Y  
467             1.0     Semiurban           Y  
477             0.0     Semiurban           N  
507             1.0         Urban           N  
576             0.0     Semiurban           N  
588             1.0     Semiurban           Y  
592             1.0     Semiurban           Y

Could we use the other columns as a clue to fill in the missing values for Gender?

In [8]:
df['Gender'].value_counts()

Male      489
Female    112
Name: Gender, dtype: int64

In [9]:
df['Dependents'].describe()

count     599
unique      4
top         0
freq      345
Name: Dependents, dtype: int64

In [10]:
#[(df['Dependents']>0) & 
df['Gender'][(df['Married']=='Yes') & (df['Dependents']!='0')].value_counts()

Male      357
Female     31
Name: Gender, dtype: int64

In [11]:
df['Gender'].value_counts()

Male      489
Female    112
Name: Gender, dtype: int64

<h2>Filling in the missing values</h2>

In [12]:
df.isnull().sum()

Loan_ID               0
Gender               13
Married               3
Dependents           15
Education             0
Self_Employed        32
ApplicantIncome       0
CoapplicantIncome     0
LoanAmount           22
Loan_Amount_Term     14
Credit_History       50
Property_Area         0
Loan_Status           0
dtype: int64

In [13]:
df['Gender'].mode()

0    Male
dtype: object

In [14]:
for i in ['Gender','Married','Dependents','Self_Employed']:
    df[i].fillna(df[i].mode()[0],inplace=True)
    
df.isnull().sum()

Loan_ID               0
Gender                0
Married               0
Dependents            0
Education             0
Self_Employed         0
ApplicantIncome       0
CoapplicantIncome     0
LoanAmount           22
Loan_Amount_Term     14
Credit_History       50
Property_Area         0
Loan_Status           0
dtype: int64

In [15]:
df['LoanAmount'].mean()

146.41216216216216

In [16]:
for i in ['LoanAmount','Loan_Amount_Term','Credit_History']:
    df[i].fillna(df[i].mean(),inplace=True)
    
df.isnull().sum()

Loan_ID              0
Gender               0
Married              0
Dependents           0
Education            0
Self_Employed        0
ApplicantIncome      0
CoapplicantIncome    0
LoanAmount           0
Loan_Amount_Term     0
Credit_History       0
Property_Area        0
Loan_Status          0
dtype: int64

In [17]:
df.head()

Loan_ID Gender Married Dependents     Education Self_Employed  \
0  LP001002   Male      No          0      Graduate            No   
1  LP001003   Male     Yes          1      Graduate            No   
2  LP001005   Male     Yes          0      Graduate           Yes   
3  LP001006   Male     Yes          0  Not Graduate            No   
4  LP001008   Male      No          0      Graduate            No   

   ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
0             5849                0.0  146.412162             360.0   
1             4583             1508.0  128.000000             360.0   
2             3000                0.0   66.000000             360.0   
3             2583             2358.0  120.000000             360.0   
4             6000                0.0  141.000000             360.0   

   Credit_History Property_Area Loan_Status  
0             1.0         Urban           Y  
1             1.0         Rural           N  
2             1.0         Urban           Y  
3             1.0         Urban           Y  
4             1.0         Urban           Y

<h2>Let's change some categorical variables to numerical variables</h2>

In [18]:
#Let us use the Label Encoder to change categorical variables to numerical
from sklearn.preprocessing import LabelEncoder

In [19]:
l_c = LabelEncoder()

In [20]:
#Let's ensure the dtypes of all the columns are uniform
df.dtypes

Loan_ID               object
Gender                object
Married               object
Dependents            object
Education             object
Self_Employed         object
ApplicantIncome        int64
CoapplicantIncome    float64
LoanAmount           float64
Loan_Amount_Term     float64
Credit_History       float64
Property_Area         object
Loan_Status           object
dtype: object

In [21]:
df['Dependents'].replace(['1','2','3+'],[1,2,3],inplace=True)
#df['Dependents'].value_counts
#df['Dependents'].astype('int')
df['Dependents'].dtype

dtype('int64')

In [22]:
df['Gender']

0        Male
1        Male
2        Male
3        Male
4        Male
        ...  
609    Female
610      Male
611      Male
612      Male
613    Female
Name: Gender, Length: 614, dtype: object

In [23]:
for i in ['Gender','Married','Dependents','Education','Self_Employed','Property_Area','Loan_Status']:
    df[i] = l_c.fit_transform(df.loc[:,i])

df.head()

Loan_ID  Gender  Married  Dependents  Education  Self_Employed  \
0  LP001002       1        0           0          0              0   
1  LP001003       1        1           1          0              0   
2  LP001005       1        1           0          0              1   
3  LP001006       1        1           0          1              0   
4  LP001008       1        0           0          0              0   

   ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
0             5849                0.0  146.412162             360.0   
1             4583             1508.0  128.000000             360.0   
2             3000                0.0   66.000000             360.0   
3             2583             2358.0  120.000000             360.0   
4             6000                0.0  141.000000             360.0   

   Credit_History  Property_Area  Loan_Status  
0             1.0              2            1  
1             1.0              0            0  
2             1.0              2            1  
3             1.0              2            1  
4             1.0              2            1

<h2>Extracting Variables and Output data</h2>

In [24]:
#Variables will be x
#Output will be y

In [25]:
Y = df['Loan_Status'].values
X = df.drop(['Loan_Status','Loan_ID'],axis=1).values
X

array([[  1.,   0.,   0., ..., 360.,   1.,   2.],
       [  1.,   1.,   1., ..., 360.,   1.,   0.],
       [  1.,   1.,   0., ..., 360.,   1.,   2.],
       ...,
       [  1.,   1.,   1., ..., 360.,   1.,   2.],
       [  1.,   1.,   2., ..., 360.,   1.,   2.],
       [  0.,   0.,   0., ..., 360.,   0.,   1.]])

In [26]:
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(X, Y, test_size=0.2, random_state = 2)

In [27]:
x_train.shape

(491, 11)

In [28]:
x_test.shape

(123, 11)

In [29]:
from sklearn.naive_bayes import GaussianNB
nb = GaussianNB()

In [30]:
nb.fit(x_train,y_train)

GaussianNB()

In [31]:
Y_pred = nb.predict(x_test)

In [32]:
from sklearn.metrics import accuracy_score
accuracy_score(Y_pred,y_test)

0.7804878048780488

In [68]:
#Let's determine the accuracy of the model on the training data itself
Y_train_pred = nb.predict(x_train)
accuracy_score(Y_train_pred,y_train)

0.8065173116089613

In [33]:
#Let's try another model (Logistic Regression)
from sklearn.linear_model import LogisticRegression
LR = LogisticRegression()

In [34]:
LR.fit(x_train,y_train)

C:\ProgramData\Anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:762: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression()

In [35]:
y_pred = LR.predict(x_test)

In [36]:
accuracy_score(y_pred, y_test)

0.7560975609756098

In [37]:
#Let's try that again
LR1 = LogisticRegression()
LR1 = LogisticRegression(C=0.01, solver='liblinear').fit(x_train,y_train)

In [38]:
y1 = LR1.predict(x_test)

In [39]:
accuracy_score(y1, y_test)

0.6747967479674797

The accuracy scores are really low. Let's try using Standard Scalar on the financial amounts.


In [40]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()

In [54]:
#scaled_X = df.drop('[LoanID]')
#for i in ['ApplicantIncome','CoapplicantIncome','LoanAmount','Loan_Amount_Term']:
scaled_X = scaler.fit_transform(X)
scaled_X

array([[ 0.47234264, -1.37208932, -0.73780632, ...,  0.27985054,
         0.45164045,  1.22329839],
       [ 0.47234264,  0.72881553,  0.25346957, ...,  0.27985054,
         0.45164045, -1.31851281],
       [ 0.47234264,  0.72881553, -0.73780632, ...,  0.27985054,
         0.45164045,  1.22329839],
       ...,
       [ 0.47234264,  0.72881553,  0.25346957, ...,  0.27985054,
         0.45164045,  1.22329839],
       [ 0.47234264,  0.72881553,  1.24474546, ...,  0.27985054,
         0.45164045,  1.22329839],
       [-2.11710719, -1.37208932, -0.73780632, ...,  0.27985054,
        -2.41044061, -0.04760721]])

In [55]:
x_train1, x_test1, y_train1, y_test1 = train_test_split(scaled_X, Y, test_size=0.2, random_state = 2)

In [57]:
#Naive Bayes again
NB = GaussianNB()

In [58]:
NB.fit(x_train1,y_train1)

GaussianNB()

In [61]:
y_scale = NB.predict(x_test1)

In [62]:
accuracy_score(y_scale, y_test1)

0.7804878048780488

In [63]:
LR1 = LogisticRegression()

In [64]:
LR1.fit(x_train1, y_train1)

LogisticRegression()

In [66]:
y2 = LR1.predict(x_test1)

In [67]:
accuracy_score(y2, y_test1)

0.7642276422764228

In [41]:
df['ApplicantIncome'].max()

81000

In [42]:
df['ApplicantIncome'].min()

150